### PR TITLE
Chore: Extract lint result cache logic (refs #9948)

### PR DESCRIPTION
**Pull request by platinumazure**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain: Extract lint result cache logic to its own utility module.

**Motivation**: This is preparation for a possible implementation of #9948, which might require that the object we cache for linted files will no longer have a 1-to-1 correspondence with the lint result object.

**What changes did you make? (Give an overview)**

Extracted cache logic for linting results (only) into its own file. The behavior should be unchanged.

**Is there anything you'd like reviewers to focus on?**

I'm not happy with the way I've set up mocks etc. in the tests. It's not immediately obvious to future contributors why some setups are in an inner `beforeEach` vs an outer or middle `before` and why more of the setups couldn't be closer together, but I couldn't figure out a great way to improve upon what I currently have. Any tips for making those more organized/easier to read (possibly with a different library)?

### CODE REVIEW

File: lib/util/lint-result-cache.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+        const fileDescriptor = this.fileEntryCache.getFileDescriptor(filePath);
+        const hashOfConfig = hashOfConfigFor(this.configHelper, filePath);
+        const changed = fileDescriptor.changed || fileDescriptor.meta.hashOfConfig !== hashOfConfig;
+
+        if (fileDescriptor.notFound || changed) {
+            return null;
+        }
+
+        // TODO (platinumazure): Hydrate any missing elements from results
```
[the remark this answers is not in the document]
Comment (platinumazure) [reply lib/util/lint-result-cache.js]: No, there's no correctness issue for now. This is a reference to where I would need to add code to implement the feature requested in #9984. I will remove this comment.

**Comment by platinumazure** [comment]
I'll rebase and then push a commit to remove the comments called out in review. Thanks!

**Comment by platinumazure** [comment]
@not-an-aardvark Friendly ping: Have I addressed your concerns?

File: tests/fixtures/lint-result-cache/test-no-errors.js [hunk 2]
```diff
@@ -0,0 +1 @@
+console.log("Hello, world!");
```
Comment (platinumazure) [inline tests/fixtures/lint-result-cache/test-no-errors.js]: I should remove this file.

**Comment by not-an-aardvark** [comment]
Sorry about the delay -- I haven't had time to look at this again, but my concerns were relatively minor, so feel free to merge if you think you've addressed them.

File: lib/cli-engine.js [hunk 3]
```diff
@@ -434,6 +419,17 @@ class CLIEngine {
         }
 
         this.config = new Config(this.options, this.linter);
+
+        if (options.cache) {
```
Comment (kaicataldo) [inline lib/cli-engine.js]: Nit: Should this maybe be `this.options.cache` for consistency?

**Comment by platinumazure** [comment]
Adding "do not merge" to avoid early merge while I rebase, delete an unnecessary file, and implement suggestions provided in feedback.

**Comment by platinumazure** [comment]
I've addressed all feedback and the Travis/Appveyor builds passed, so I'm going to merge. Thanks to everyone who reviewed for your feedback!

**Pull request merged**